matrix-client.el

```diff
@@ -267,22 +281,27 @@ Otherwise, use the room name or alias."
 If [`matrix-client-use-auth-source'] is non-nil, attempt to log
 in using data from auth-source.  Otherwise, prompt for username
 and password."
-  (let* ((auth-source-creation-prompts (a-list 'username "Matrix identity: "
-                                               'secret "Matrix password for %u (homeserver: %h): "))
-         (found (nth 0 (auth-source-search :max 1
-                                           :host (oref con :base-url)
-                                           :user username
-                                           :require '(:user :secret)
-                                           :create t))))
-    (when (and found
-               (matrix-login-with-password con (plist-get found :user)
-                                           (let ((secret (plist-get found :secret)))
-                                             (if (functionp secret)
-                                                 (funcall secret)
-                                               secret))))
-      (oset con :username (plist-get found :user))
-      (when-let ((save-func (plist-get found :save-function)))
-        (funcall save-func)))))
+  (let* ((username (or username (read-from-minibuffer "Username: ")))
+         (password (read-passwd "Password: ")))
+    (when (matrix-login-with-password con username password)
+      (oset con :username username)
+      (when matrix-client-save-token
+        (matrix-client-save-token con)))))
+
+(cl-defmethod matrix-client-save-token ((con matrix-client-connection))
+  "Save username and access token for connection CON to file."
+  (with-temp-file matrix-client-save-token-file
+    (with-slots (username token) con
+      (prin1 (a-list 'username username
+                     'token token)
+             (current-buffer))))
+  ;; Ensure permissions are safe
+  (chmod matrix-client-save-token-file #o600))
+
+(defun matrix-client-load-token ()
+  "Return saved username and access token from file."
+  (when (f-exists? matrix-client-save-token-file)
```

Comment:
You need to add a `require` and depdendency for `f.el` here. `f-exists?` isnt in melpa stable though (?!?) but I think that's fine.

Reply:
Good catch, I forgot to require `f`.  Thanks.